galsim/wfirst/wfirst_wcs.py

```diff
@@ -0,0 +1,478 @@
+    (+f2) is at a position angle pa_fpa = pa_obsy + 90 + theta_fpa.
+
+    @param PA        Position angle of the observatory Y axis, unless `PA_is_FPA=True`, in which
+                     case it's the position angle of the FPA.  Must be provided as a galsim.Angle.
+    @param world_pos A galsim.CelestialCoord indicating the center of the FPA.
+    @param PA_is_FPA If True, then the position angle that was provided was the PA of the focal
+                     plane array, not the observatory. [default: False]
+    @param SCAs      A single number or iterable giving the SCAs for which the WCS should be
+                     obtained.  If None, then the WCS is calculated for all SCAs.
+                     [default: None]
+    @returns a dict of WCS objects for each SCA.
+    """
+    # Check which SCAs are to be done using a helper routine in this module.
+    SCAs = galsim.wfirst._parse_SCAs(SCAs)
+
+    # Enforce type for PA
+    if not isinstance(PA, galsim.Angle):
+        raise TypeError("Position angle must be a galsim.Angle!")
+
+    # Parse input position
+    if not isinstance(world_pos, galsim.CelestialCoord):
+        raise TypeError("Position on the sky must be given as a galsim.CelestialCoord!")
+
+    # Compute position angle of FPA f2 axis, where positive corresponds to the angle east of North.
+    if PA_is_FPA:
+        pa_fpa = PA
+        pa_obsy = PA - 90.*galsim.degrees - theta_fpa
+    else:
+        pa_obsy = PA
+        pa_fpa = PA + 90.*galsim.degrees + theta_fpa
+    cos_pa = np.cos(pa_fpa.rad())
+    sin_pa = np.sin(pa_fpa.rad())
+
+    # Figure out tangent-plane positions for FPA center:
+    xc_fpa_tp, yc_fpa_tp = _det_to_tangplane_positions(xc_fpa, yc_fpa)
+
+    # Note, this routine reads in the coeffs.  We don't use them until later, but read them in for
+    # all SCAs at once.
+    a_sip, b_sip = _parse_sip_file(sip_filename)
+
+    # Loop over SCAs:
+    wcs_dict = {}
+    for i_sca in SCAs:
+        # Set up the header.
+        header = galsim.FitsHeader()
+        # Populate some necessary variables in the FITS header that are always the same, regardless of
+        # input and SCA number.
+        _populate_required_fields(header)
+
+        # And populate some things that just depend on the overall locations or other input, not on
+        # the SCA.
+        header['RA_TARG'] = (world_pos.ra / galsim.degrees,
+                             "right ascension of the target (deg) (J2000)")
+        header['DEC_TARG'] = (world_pos.dec / galsim.degrees,
+                              "declination of the target (deg) (J2000)")
+        header['PA_OBSY'] = (pa_obsy / galsim.degrees, "position angle of observatory Y axis (deg)")
+        header['PA_FPA'] = (pa_fpa / galsim.degrees, "position angle of FPA Y axis (deg)")
+
+        # Finally do all the SCA-specific stuff.
+        header['SCA_NUM'] = (i_sca, "SCA number (1 - 18)")
```

Comment:
My original idea was that the user would need to make an `hduList` and write to an hdu instead of a file.  Then have a function that would add items to the header:
```
im = galsim.Image(...)
im.wcs = wcs  # which implies some header keys already
# ... make image
hdu_list = pyfits.HDUList()
im.write(hdu_list = hdu_list)
galsim.wfirst.populateFitsHeader(hdu_list[-1].header, ...)
galsim.fits.writeFile(file_name, hdu_list)
```

But I think I may have a better idea.  How's this:
```
im = galsim.Image(...)
im.wcs = wcs
# ... make image
im.header = galsim.wfirst.makeFitsHeader(...)
im.write(file_name)
```
Then the `write` function could look for a header object in image, and if there is one there, it uses that.  Otherwise it does the normal thing and lets the wcs write itself to the header.  This seems like a pretty natural syntax I think.

I'm not sure whether the wcs should also be written to the header if the user provides a header.  And which order if so. I can think of three plausible options:
1. Start by writing the wcs to the header, and then update with whatever else is in `image.header`, possibly overwriting things that the wcs specified
2. Start with `image.header`, and then let the wcs write itself to make sure that's all correct.
3. Just use `image.header` and not call the `wcs.writeToFitsHeader` function if the user has provided a header.  i.e. the user is responsible for using that function if they want it in the final header.